### PR TITLE
[BEAM-4000] Futurize io subpackage

### PR DESCRIPTION
**Pull request by Fematich**
This pull request prepares the io subpackage for Python 3 support. This PR is part of a series in which all subpackages will be updated using the same approach.
This approach has been documented [here ](https://docs.google.com/document/d/1xDG0MWVlDKDPu_IW9gtMvxi2S9I0GB0VDTkPhjXT0nE/edit) and the first pull request in the series (Futurize coders subpackage) demonstrating this approach can be found at #5053.

R: @aaltay @tvalentyn @RobbeSneyders @charlesccychen

### CODE REVIEW

File: sdks/python/apache_beam/io/avroio.py [hunk 1]
```diff
@@ -60,6 +63,8 @@
 from apache_beam.io.iobase import Read
 from apache_beam.transforms import PTransform
 
+standard_library.install_aliases()
```
Comment (superbobry) [inline sdks/python/apache_beam/io/avroio.py]: Is this needed?

File: sdks/python/apache_beam/io/filebasedsink.py [hunk 2]
```diff
@@ -307,7 +314,7 @@ def _rename_batch(batch):
           return exceptions
 
       exception_batches = util.run_using_threadpool(
-          _rename_batch, zip(source_file_batch, destination_file_batch),
+          _rename_batch, list(zip(source_file_batch, destination_file_batch)),
```
Comment (superbobry) [inline sdks/python/apache_beam/io/filebasedsink.py]: IIUC run_using_threadpool accepts any iterable, so forcing a list is redundant.

File: sdks/python/apache_beam/io/filebasedsource.py [hunk 3]
```diff
@@ -49,6 +48,13 @@
 
 __all__ = ['FileBasedSource']
 
+try:
+  unicode       # pylint: disable=unicode-builtin
```
Comment (superbobry) [inline sdks/python/apache_beam/io/filebasedsource.py]: How about 

```python
from past.builtins import long, unicode
```

This applies to other similar try-except imports in this (and related) PRs.

File: sdks/python/apache_beam/io/filebasedsource.py [hunk 4]
```diff
@@ -235,11 +241,11 @@ class _SingleFileSource(iobase.BoundedSource):
 
   def __init__(self, file_based_source, file_name, start_offset, stop_offset,
                min_bundle_size=0, splittable=True):
-    if not isinstance(start_offset, integer_types):
+    if not isinstance(start_offset, (int, long)):
```
Comment (superbobry) [inline sdks/python/apache_beam/io/filebasedsource.py]: If you import from `past` you can replace this with just `int`.

See http://python-future.org/compatible_idioms.html#long-integers

File: sdks/python/apache_beam/io/filesystem.py [hunk 5]
```diff
@@ -131,7 +143,7 @@ def __init__(self,
 
     if self.readable():
       self._read_size = read_size
-      self._read_buffer = cStringIO.StringIO()
+      self._read_buffer = BytesIO()
```
Comment (superbobry) [inline sdks/python/apache_beam/io/filesystem.py]: Consistency nitpick: other modules use `io` in a qualified way.

File: sdks/python/apache_beam/io/filesystems_test.py [hunk 6]
```diff
@@ -124,7 +126,7 @@ def test_match_file_exception(self):
     with self.assertRaisesRegexp(BeamIOError,
                                  r'^Unable to get the Filesystem') as error:
       FileSystems.match([None])
-    self.assertEqual(error.exception.exception_details.keys(), [None])
+    self.assertEqual(list(error.exception.exception_details.keys()), [None])
```
Comment (superbobry) [inline sdks/python/apache_beam/io/filesystems_test.py]: Consider using list(d) instead of list(d.keys()).

File: sdks/python/apache_beam/io/gcp/datastore/v1/helper.py [hunk 7]
```diff
@@ -295,6 +300,9 @@ def _next_batch(self):
     resp = self._datastore.run_query(self._req)
     return resp
 
+  def __next__(self):
+    return next(self.__iter__())
```
Comment (superbobry) [inline sdks/python/apache_beam/io/gcp/datastore/v1/helper.py]: This will always return the first element. Is this the expected behaviour?

File: sdks/python/apache_beam/io/gcp/datastore/v1/helper_test.py [hunk 8]
```diff
@@ -82,7 +85,7 @@ def test_query_iterator(self):
         self.permanent_retriable_datastore_failure)
     query_iterator = helper.QueryIterator("project", None, self._query,
                                           self._mock_datastore)
-    self.assertRaises(RPCError, iter(query_iterator).next)
+    self.assertRaises(RPCError, query_iterator.__next__)
```
Comment (superbobry) [inline sdks/python/apache_beam/io/gcp/datastore/v1/helper_test.py]: This will not work on Python2. How about

```python
from future.builtins import next

self.assertRaises(..., lambda: next(query_iterator))
```

File: sdks/python/apache_beam/io/gcp/gcsio.py [hunk 9]
```diff
@@ -483,6 +489,7 @@ def size(self):
     return self._size
 
   def get_range(self, start, end):
+    self._download_stream.seek(0)
```
Comment (superbobry) [inline sdks/python/apache_beam/io/gcp/gcsio.py]: Why is this needed in Python 3?

File: sdks/python/apache_beam/io/gcp/pubsub.py [hunk 10]
```diff
@@ -49,6 +48,10 @@
 except ImportError:
   pubsub_pb2 = None
 
+try:
+  basestring
```
Comment (superbobry) [inline sdks/python/apache_beam/io/gcp/pubsub.py]:
```python
from past.builtins import basestring
```

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
Comment (charlesccychen) [reply sdks/python/apache_beam/io/avroio.py]: This is needed for `import io`, but is misplaced.

File: sdks/python/apache_beam/io/filebasedsink.py (hunk 2 above)
Comment (charlesccychen) [reply sdks/python/apache_beam/io/filebasedsink.py]: Reading `util.run_using_threadpool`, it looks like the code calls `len(inputs)`, which may not be compatible with all iterables.

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
Comment (superbobry) [reply sdks/python/apache_beam/io/avroio.py]: `io` is [available](https://docs.python.org/2/library/io.html) starting from Python 2.6.

File: sdks/python/apache_beam/io/filebasedsink.py (hunk 2 above)
Comment (superbobry) [reply sdks/python/apache_beam/io/filebasedsink.py]: Ack, please discard my initial comment.

File: sdks/python/apache_beam/io/filebasedsource.py (hunk 3 above)
Comment (Fematich) [reply sdks/python/apache_beam/io/filebasedsource.py]: I used the try/except method to be consistent with the methodology implemented in #5053 and outlined in [the doc](https://docs.google.com/document/d/1xDG0MWVlDKDPu_IW9gtMvxi2S9I0GB0VDTkPhjXT0nE/edit#heading=h.ir03vqq4rotn) by @RobbeSneyders. I believe there where some issues in the coder package for the typechecks using `past.utils`

Comment (superbobry) [reply sdks/python/apache_beam/io/filebasedsource.py]: I'd humbly suggest reconsidering (unless there are indeed issues with using `past` in some modules) as try-except is much more verbose syntactically.

File: sdks/python/apache_beam/io/gcp/gcsio.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Fematich) [reply sdks/python/apache_beam/io/gcp/gcsio.py]: Yes, this is a difference in behavior between cStringIO and BytesIO. In BytesIO doesn't move the file pointer, see [here](https://stackoverflow.com/questions/39109068/why-does-truncating-a-bytesio-mess-it-up)

File: sdks/python/apache_beam/io/filebasedsource.py (hunk 3 above)
Comment (RobbeSneyders) [reply sdks/python/apache_beam/io/filebasedsource.py]: The try/except was indeed used to prevent problems with typechecking when using `builtins` from the future package.
However, I tested this again, and it seems that there are no such problems using the `past.builtins`.

Comment (RobbeSneyders) [reply sdks/python/apache_beam/io/filebasedsource.py]: I have created the issue [BEAM-4730](https://issues.apache.org/jira/browse/BEAM-4730) to apply this change to already ported packages.

Comment (superbobry) [reply sdks/python/apache_beam/io/filebasedsource.py]: This is done in #5869.

Comment (RobbeSneyders) [reply sdks/python/apache_beam/io/filebasedsource.py]: Great, thank you!

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
Comment (Fematich) [reply sdks/python/apache_beam/io/avroio.py]: Removed

File: sdks/python/apache_beam/io/gcp/datastore/v1/helper.py (hunk 7 above)
Comment (Fematich) [reply sdks/python/apache_beam/io/gcp/datastore/v1/helper.py]: No this is indeed an error. This method is actually now also superfluous if I use your suggestion for the test in https://github.com/apache/beam/pull/5715#discussion_r199078083. I removed it

File: sdks/python/apache_beam/io/gcp/gcsio.py (hunk 9 above)
Comment (Fematich) [reply sdks/python/apache_beam/io/gcp/gcsio.py]: Difference in behavior between cStringIO and BytesIO. In BytesIO truncate doesn't move the file pointer, see [here](https://stackoverflow.com/questions/39109068/why-does-truncating-a-bytesio-mess-it-up)

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Fematich) [reply sdks/python/apache_beam/io/avroio.py]: Order should be OK, see https://github.com/apache/beam/pull/5373#discussion_r200076690

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 11]
```diff
@@ -44,6 +48,8 @@
 from apache_beam.transforms.display import DisplayData
 from apache_beam.transforms.display_test import DisplayDataItemMatcher
 
+standard_library.install_aliases()
```
[the remark this answers is not in the document]
Comment (Fematich) [reply sdks/python/apache_beam/io/filebasedsource_test.py]: Order should be OK, see https://github.com/apache/beam/pull/5373#discussion_r200076690

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
Comment (Fematich) [reply sdks/python/apache_beam/io/avroio.py]: I removed all the `install_aliases` caused by `io`

File: sdks/python/apache_beam/io/gcp/gcsio.py (hunk 9 above)
Comment (superbobry) [reply sdks/python/apache_beam/io/gcp/gcsio.py]: Yes, thanks for the clarification.

File: sdks/python/apache_beam/io/gcp/pubsub.py [hunk 12]
```diff
@@ -33,8 +33,9 @@
 from __future__ import absolute_import
 
 import re
+from builtins import object
 
-from six import text_type
+from past.builtins import basestring
```
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: It seems that on Python 2 this change translates to changing the `element_type` type from `unicode` to `bytes`/`str` in line 176. 
@chamikaramj Do you have an intuition if this could bite us?
CC: @altay

[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: Correction: `past.builtins.basestring` is configured to mimic a superclass of both `str` and `unicode` in Python 2. So possibly there is no significant change in behavior but I am not 100% sure.

[the remark this answers is not in the document]
Comment (Fematich) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: It's indeed a superclass of both `str` and `unicode`.  Only in https://github.com/apache/beam/pull/5373#discussion_r195623409 I observed the real need to use this superclass. However for this example, it seems like `past.builtins.unicode` would be a more compatible replacement (https://pythonhosted.org/six/#six.text_type) since I can't recall an explicit reason to use basestring instead of unicode here.

[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: https://github.com/apache/beam/pull/6144 cleans this up here as well as in the place you mentioned, since we have reverted `# cython: language_level=3`.
